Sources/Smithy/Context.swift

```diff
@@ -5,46 +5,63 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
+import class Foundation.NSRecursiveLock
+
 public class Context {
-    public var attributes: Attributes
+    private var _attributes: Attributes
+    private let lock = NSRecursiveLock()
 
     public init(attributes: Attributes) {
-        self.attributes = attributes
+        self._attributes = attributes
     }
 
     public func toBuilder() -> ContextBuilder {
-        let builder = ContextBuilder()
-        builder.attributes = self.attributes
-        return builder
+        ContextBuilder(attributes: accessAttributes())
     }
 
     public func getLogger() -> LogAgent? {
-        return attributes.get(key: AttributeKeys.logger)
+        return accessAttributes().get(key: AttributeKeys.logger)
+    }
+
+    @discardableResult
+    private func accessAttributes(accessor: ((inout Attributes) -> Void)? = nil) -> Attributes {
+        lock.lock()
+        defer { lock.unlock() }
+        accessor?(&_attributes)
+        return _attributes
```

Comment:
Question: I am not sure if this is fully thread safe. If I am reading this correctly it looks like the thing that happens within scope of the lock is making a copy of _attributes. Then when get is called via accessAttributes().get(key: key) accessAttributes() creates a copy of _attributes under a lock, the lock is released, and then .get(key: key) is called without any lock. I think to make this thread safe we need the operations themselves to be performed under a lock -- lmk what you think

Reply:
That operation (i.e. what happens in `getLogger()`) is safe because Attributes is a value type, i.e. it is copied when a new reference to it is created.  The copy of the attributes that is returned by `accessAttributes()` is not referenced anywhere else, so it is safe to access the `logger` (or any other attribute).

You would not want to modify that copy of attributes to modify the context, because you're modifying a copy, the change will not be reflected in the Context the attributes were copied from.

Reply:
Honestly that getLogger() method can be simplified down to just `get(AttributeKeys.logger)`.  Same effect, same thread safety, simpler code.